Initial support for pthreads with wasm

dschuff opened the pull request:
Also adds support for testing with both wasm and JS+SAB, but there's no clean way to enable it yet.
Tested with `EMSCRIPTEN_BROWSER='/path/to/chrome --user-data-dir=clean_dir --js-flags=--experimental-wasm-threads --enable-features=WebAssembly --disable-features=WebAssemblyTrapHandler'` with a locally-built chrome but it should work with canary also.

dschuff commented on src/preamble.js:
```diff
@@ -1839,7 +1849,7 @@ function addRunDependency(id) {
   // We should never get here in pthreads (could no-op this out if called in pthreads, but that might indicate a bug in caller side,
   // so good to be very explicit)
-  assert(!ENVIRONMENT_IS_PTHREAD);
+  assert(!ENVIRONMENT_IS_PTHREAD || id === 'wasm-instantiate');
```
This is uuuuugly.

juj replied:
Did you make this change in order to get async instantiation working? It would probably be better if we avoided this altogether and follow the path `"We can just use sync instantiation in the worker."` that you wrote below?

dschuff replied:
Actually that's not quite right; preamble.js adds the dependency regardless of whether a custom instantiator is used. We can bypass that in threads (see 9d9762) which I guess is a little bit nicer than this. I also noticed that right now the user can't override wasm-instantiate in a pthread right now (only on the main thread). Is that something we need?

kripken commented on src/preamble.js:
```diff
@@ -1237,6 +1238,15 @@ if (typeof Atomics === 'undefined') {
 }
 
+#else
+// TODO: Something smart about BINARYEN_MEM_MAX. Some default, also rename it.
```
is it right that wasm can't support memory growth with shared memory? if so, we can just show an error if growth or max > initial?

dschuff replied:
It can support growth, just not unconstrained growth. Or in other words, you can have max > initial, but you can't set an initial without a max.

kripken replied:
I see, ok.

I still don't understand this code, though. Why do we create the memory here instead of later down where we create it normally (which also handles MEM_MAX)?

dschuff replied:
Basically just because of the overall structure of the ifdefs. All the pthreads stuff is first, and everything below is in the else (not pthreads). This whole nest of code is really ugly and complex because of all the options, should probably cleaned up and/or the options rethought.

juj replied:
Need to resolve this comment - I think we can make command line requirement to pass `BINARYEN_MEM_MAX` so that frontend makes sure in multithreaded mode this gets set?

dschuff replied:
OK, made it a requirement to add `WASM_MEM_MAX` when using wasm+pthreads and memory growth. I tried to add a test but it doesn't look like memory growth on a thread is working right now. Will address that separately.

kripken commented on src/preamble.js:
```diff
@@ -2303,6 +2313,10 @@ function integrateWasmJS() {
 #endif
+#if USE_PTHREADS
+      // Keeep a reference to the compiled module so we can post it to the workers.
+      Module['wasmModule'] = output['module'];
```
the module should be saved as `Module['asm']` already

juj replied:
The file `preamble.js` has `Module['asm'] = exports;`, so `'asm'` seems to be getting the exports object, rather than the Module. So it looks like we do need to store this manually(?)

I think what we'd like is to make `receiveInstance()` take a second parameter that is the module, so that applications that instantiate themselves will also get the module stored for threads.

I do hate this a bit that we do have to store the module to be able to instantiate a new thread - that wastes memory; and I was pondering if we could reuse IndexedDB and the thread pool idea to optimize, i.e. if application knows they only need fixed K threads at startup, the could precreate a pool of that size to be able to throw away the Module afterwards. If new threads want to spawn after that, they would take a slow path of reading the Module from IndexedDB for instantiation. Although not sure if that's worth the trouble to optimize, so let's leave that for later.

kripken commented on tests/test_browser.py:
```diff
@@ -3108,19 +3108,26 @@ def prep_no_SAB(self):
     '''))
 
+  # Run a browser test both with and without wasm
+  def btest_wasm(self, *args, **kwargs):
```
for running as both asm and wasm, see the `also_proxied` flag to btest which makes us test both regular and proxied mode. maybe we can do the same pattern for asm and wasm.

kripken replied:
I still don't see why we need a new method `btest_wasm`, can't be do something parallel to `also_proxied`, like `also_wasm`?

otherwise lgtm

kripken replied:
Or is this different because it is off for now?

dschuff replied:
No, sorry I just hadn't gotten to that yet. Its switched now. I don't see the feature-detection mechanism you mentioned though. Since the tests run in a random order we can't just rely on the outcome of one of them to gate the rest. I thought I might just add a flag to runner.py for now.

dschuff replied:
BTW, chrome needs a flag to enable threads (which means modifying `EMSCRIPTEN_BROWSER` in the env) so I thought maybe just another env var for testing it since we have an env var anyway. Does nightly need a separate flag, or is it just on by default there?

kripken replied:
Not sure about how we avoid pthreads tests if no pthreads support - @juj?

Also @juj would know about firefox nightly needing a flag for wasm threads or not as he tested it.

kripken commented:
Need to think about the testing here. I think we have a mechanism to check if pthreads is supported and we skip those tests otherwise, maybe we need something similar for wasm-pthreads, since right now they aren't in firefox, so those tests would break.

kripken commented:
There is now a firefox build with wasm threads support. I tested a few simple things and didn't see any issues. I did need to add
````
+    if shared.Settings.USE_PTHREADS:
+      cmd += ['--enable-threads']
````
around line 2263 or emcc.py though.

juj commented:
Perfect, thanks @dschuff for getting this going!

I'm testing this against the Multithreading series of branches, which I'm looking to resume landing shortly. Currently running test suites on Firefox Nightly, so I'll get back with feedback on this in a moment.

juj commented:
Test suites pass in FF Nightly, and also some code compiles; The tree https://github.com/juj/emscripten/commits/multithreading_5 is the current "latest state" for multithreading support that the Multithreading PRs are landing towards, see in particular https://github.com/juj/emscripten/commit/a1cf5209c691fe40bd69f9ed22871f205b3b0bf2 and https://github.com/juj/emscripten/commit/8690b9a73bdb4ade048b21c0c76a90aba09654d7 that relate to this PR.

Unfortunately it looks like real world(tm) codebases don't still compile, I'm currently hunting some odd codegen related bug to understand this better. It could be related to the global data section initialization commit above, though not completely sure.

kripken commented:
lgtm. the new wasm-pthreads are off by default, looking for that env var, correct?

@juj should look at this code too (before or after landing is ok i think).

juj commented on src/preamble.js:
```diff
@@ -2264,15 +2274,22 @@ function integrateWasmJS() {
+      // Keep a reference to the compiled module so we can post it to the workers.
+      Module['wasmModule'] = module;
+#endif
+      if(!ENVIRONMENT_IS_PTHREAD) {
```
`ENVIRONMENT_IS_PTHREAD` does not exist when not building with `-s USE_PTHREADS=1` mode, so this should be gated inside `if USE_PTHREADS`.

dschuff replied:
This change (https://github.com/kripken/emscripten/pull/5710/files#diff-8f35bc569987bab9a3691d986d90e75fR72) makes it exist unconditionally (just like the rest of the `ENVIRONMENT_IS_FOO` checks).

juj commented on src/preamble.js:
```diff
@@ -2264,15 +2274,22 @@ function integrateWasmJS() {
+      }
+    }
+    if (!ENVIRONMENT_IS_PTHREAD) {
+      addRunDependency('wasm-instantiate'); // we can't run yet (except in a pthread, where we have a custom sync instantiator)
```
Same here

dschuff commented:
The compiler will do what you expect if you use `-s WASM=1 -s USE_PTHREADS=1` (previously it would assert). But the tests are gated on the env var.

juj commented on src/pthread-main.js:
```diff
@@ -52,19 +52,39 @@ Module['print'] = threadPrint;
+  // We can just use sync instantiation in the worker.
+  instance = new WebAssembly.Instance(Module['wasmModule'], info);
+  // We don't need the module anymore; new threads will be spawned from the main thread.
+  // XXX is this true?
```
Yes, this is true, this comment can be removed. We don't want to chain Worker creations main thread -> Worker -> Worker, since we don't want to have deep chains, and main browser thread will then be able to message to all of them. So indeed in Worker the `Module['wasmModule']` bit is no longer needed. I'd prefer `delete Module['wasmModule']` over `Module['wasmModule'] = undefined;` however.

juj commented on src/pthread-main.js:
```diff
@@ -52,19 +52,39 @@ Module['print'] = threadPrint;
+  Module['wasmModule'] = undefined;
+  receiveInstance(instance);
+  return instance.exports;
+}
```
The whole block above should be `#if WASM` - there is still a short window that I want to retain both asm.js and wasm multithreading, so let's gate this accordingly.

juj commented on src/pthread-main.js:
```diff
@@ -52,19 +52,39 @@ Module['print'] = threadPrint;
       DYNAMICTOP_PTR = e.data.DYNAMICTOP_PTR;
 
+
+      if (e.data.wasmModule) {
```
Let's make this a `#if WASM` as well, because in Wasm mode, `wasmModule` is required to exist - if it doesn't we are in trouble, so in `#if WASM` we cannot proceed if `e.data.wasmModule` did not exist.

juj commented on src/pthread-main.js:
```diff
@@ -52,19 +52,39 @@ Module['print'] = threadPrint;
+        Module['wasmMemory'] = e.data.wasmMemory;
+        buffer = Module['wasmMemory'].buffer;
+      } else {
+        buffer = e.data.buffer;
```
This would be `#if !WASM`, or an `#else` block to above.

juj commented on src/shell.js:
```diff
@@ -69,6 +69,8 @@ if (!ENVIRONMENT_IS_PTHREAD) ENVIRONMENT_IS_PTHREAD = false; // ENVIRONMENT_IS_P
 if (!ENVIRONMENT_IS_PTHREAD) PthreadWorkerInit = {};
 var currentScriptUrl = (typeof document !== 'undefined' && document.currentScript) ? document.currentScript.src : undefined;
+#else
+var ENVIRONMENT_IS_PTHREAD = false;
```
Let's not add this in at all, in singlethreaded builds we want to never see a variable `ENVIRONMENT_IS_PTHREAD` anyhwere. I like to keep this strict to enforce a "don't pay for what you don't use" barrier on this feature.

kripken commented:
I had an idea, instead of adding `, also_wasm=True` to all those tests, could we apply that behavior if an env var is set? Then to run these tests, we'd do something like `EMCC_BROWSER_ALSO_WASM=1 ./tests/runner.py browser.test_pthread*`

dschuff commented:
Sorry, by "on by default" I meant the `btest` function's `also_wasm` argument defaults to True (thus allowing callers to set it to False), but also gate running any of the wasm thread tests behind an env var.

kripken commented:
In that case, it sounds like setting `also_wasm` to False is similar to the `@no_wasm` decorator?

dschuff commented:
It's similar. With the core tests we run the whole suite in a mode (asm2g, binaryen0, etc) which could be wasm or not, (and the flag is handled separately from the tests) and @no_wasm` skips the entire test case if it is. Currently the browser test system doesn't have the same notion of modes, so the test cases just exercise all the flags they want. Part of the reason is that the core tests want to exercise different compiler modes across a whole bunch of tests, whereas the browser tests are mostly trying to exercise the runtime, which is less affected by the compiler modes Therefore we don't want to run the whole suite across all modes because it would be redundant and slow.  We could try to make the browser test model more like core; whether that makes sense probably depends on how much the things that the suite wants to exercise vary across the modes. My intuition is that it wouldn't be worth it. btw threads are a notable exception because they can't currently be tested in the shell. Really we would want to put most of those tests in core (if there were a good way to use workers in shells/node).

kripken commented:
Fair enough, we can leave test improvements for some other time.

lars-t-hansen commented:
>  btw threads are a notable exception because they can't currently be tested in the shell. Really we would want to put most of those tests in core (if there were a good way to use workers in shells/node).

Working on it: https://github.com/WebAssembly/threads/issues/52

dschuff commented:
OK I think i've addressed all the comments and have checked that nothing existing breaks. Shoudl be good, but I'll wait for CI. The circleci build seems to have completed without doing anything; is there something I need to do to make that work?

kripken commented:
CircleCI will work once you get the changes on incoming, but not worth merging incoming into here for that, as they overlap the travis tests now. (Although they are faster, so it might actually be worth it ;)

Please run the browser tests locally, though, as they don't run on the bots.

dschuff commented:
OK, verified no new breakage.

kripken commented:
Great!

Do you need to keep the branch around, or can it be deleted?

dschuff commented:
Deleted.
